src/shared/ui/picker.ts

```diff
@@ -145,3 +148,213 @@ export function verifySinglePickerOutput<T extends vscode.QuickPickItem>(choices
+// TODO: Add manual pagination?
+/**
+ * Grants a `vscode.QuickPick` iteration capabilities:
+ * * Pause/unpause loading
+ * * Refresh capabilities
+ * * No item/error nodes
+ *
+ * External control is still done via `picker.promptUser` as if it were a normal QuickPick.
+ */
+export class IteratingQuickPickController<TResponse> {
+    private isDone: boolean = false
+    private isPaused: boolean = false
+    private iterator: AsyncIterator<vscode.QuickPickItem[]>
+    private activeIterationTime: Date
+
+    // Default constructs are public static so they can be validated aganist by other functions.
+    public static readonly REFRESH_BUTTON: vscode.QuickInputButton = {
+        iconPath: new vscode.ThemeIcon('refresh'),
+        tooltip: localize('AWS.generic.refresh', 'Refresh'),
+    }
+    public static readonly NO_ITEMS_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.noItemsFound.label', '[No items found]'),
+        detail: localize('AWS.picker.dynamic.noItemsFound.detail', 'Click here to go back'),
+        alwaysShow: true,
+    }
+    public static readonly ERROR_ITEM: vscode.QuickPickItem = {
+        label: localize('AWS.picker.dynamic.errorNode.label', 'There was an error retrieving more items.'),
+        alwaysShow: true,
+    }
+
+    /**
+     * @param quickPick A `vscode.QuickPick` to grant iterating capabilities
+     * @param populator An IteratingQuickPickPopulator which can call an iterator and return QuickPickItems.
+     */
+    public constructor(
+        private readonly quickPick: vscode.QuickPick<vscode.QuickPickItem>,
+        private readonly populator: IteratingQuickPickPopulator<TResponse>
+    ) {
+        // append buttons specific to iterating quickPick
+        this.quickPick.buttons = [...this.quickPick.buttons, IteratingQuickPickController.REFRESH_BUTTON]
+
+        this.iterator = this.populator.getPickIterator()
+
+        this.activeIterationTime = new Date()
+    }
+
+    /**
+     * Pauses any existing item loading.
+     * Call upon selecting an item in order to pause additional background loading.
+     */
+    public pauseRequests(): void {
+        this.isPaused = true
+    }
+
+    /**
+     * Starts item loading. Restarts from existing state if paused.
+     * Useful for manual pagination, losing quick pick focus (e.g. via external link), throttling, etc.
+     */
+    public startRequests(): void {
+        this.loadItems(false)
+    }
+
+    /**
+     * Resets quick pick's state. Useful for manual refreshes with the same query parameters
+     */
+    public reset(): void {
+        this.loadItems(true)
+    }
+
+    private async loadItems(reset: boolean): Promise<void> {
+        if (reset) {
+            this.activeIterationTime = new Date()
+            this.isDone = false
+            this.populator.reset()
+            this.iterator = this.populator.getPickIterator()
+        }
+        const scopeIterationTime = this.activeIterationTime
```

Comment:
Would consider using `AsyncLock` for this. It's done a few times in the code base and used for the s3 concurrency.

https://github.com/aws/aws-toolkit-vscode/commit/b8cc9af36708b63583abf627c3a8133324c87f41#diff-570463423bcae6a2dad31bae66461345R85-R103

All it does is queues up concurrent requests and chains them to the end of the promise with `.then` to make sure they run one after another.

You can make sure to only start a new loop once the last one has been halted (via pause or done or reset)

Reply:
I'm not 100% sure if this is the behavior I want. Correct me if I'm wrong, but this is for not showing a duplicate listing while clicking the pagination button, correct? In this case, the loop handles the pagination. The await and the iterator will control the next token. 

Instead, this timestamp is for starting multiple new iterations, e.g. hitting the refresh button 10 times in a second. This way, we don't keep flashing multiple versions of the list, we only show the most recent list. This also protects appending artifacts from the last refresh cycle to the current list.

Reply:
Looks like async-lock isn't necessary with the promise model I moved to. It could be nice for cutting down on network calls but breaking the lock can be tricky (especially if a user clicks out of the quick pick or selects an item)